dynamic_instrumentation/src/main.rs

```diff
@@ -0,0 +1,413 @@
+            .collect()
+    }
+
+    pub fn from_session_options(
+        session_options: Options,
+        free_matches: &[String],
+    ) -> eyre::Result<Self> {
+        let src_path = match free_matches {
+            [src_path] => {
+                if src_path == "-" {
+                    // We don't actually ever read from this, so I think this is a fine translation to a real [`Path`].
+                    "/dev/stdin".into()
+                } else {
+                    fs_err::canonicalize(src_path)?
+                }
+            }
+            free_matches => {
+                return Err(eyre!(
+                    "`rustc` args `matches.free` is not a single source path: {free_matches:?}"
+                ))
+            }
+        };
+        let Options {
+            crate_name,
+            crate_types,
+            edition,
+            ..
+        } = session_options;
+        let crate_name = crate_name.ok_or_else(|| eyre!("no crate_name specified by `cargo`"))?;
+        Ok(Self {
+            crate_name: crate_name.as_str().into(),
+            src_path,
+            crate_types: crate_types.into_iter().map(CrateType::from).collect(),
+            edition: edition.into(),
+        })
+    }
+
+    pub fn from_metadata(metadata: &Metadata) -> eyre::Result<Vec<Self>> {
+        Self::from_metadata_package(
+            metadata
+                .root_package()
+                .ok_or_else(|| eyre!("no root package found by `cargo`"))?,
+        )
+    }
+
+    pub fn from_rustc_args(at_args: &[String]) -> eyre::Result<Self> {
+        let args = rustc_driver::args::arg_expand_all(at_args);
+        let matches = rustc_driver::handle_options(&args)
+            .ok_or_else(|| eyre!("failed to parse `rustc` args"))?;
+        let session_options = rustc_session::config::build_session_options(&matches);
+        let crate_target = CrateTarget::from_session_options(session_options, &matches.free)?;
+        Ok(crate_target)
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+struct InstrumentInfo {
+    crate_targets: Vec<CrateTarget>,
+    metadata: PathBuf,
+}
+
+fn get_sysroot_fast() -> Option<PathBuf> {
+    let sysroot = [
+        env::var_os("RUSTUP_HOME")?,
+        "toolchains".into(),
+        env::var_os("RUSTUP_TOOLCHAIN")?,
+    ]
+    .into_iter()
+    .collect();
+    Some(sysroot)
+}
+
+fn get_sysroot_slow() -> eyre::Result<PathBuf> {
+    todo!("use `rustc --print sysroot` to support non-`rustup` cases, which @fw-immunant uses")
+}
+
+fn get_sysroot() -> eyre::Result<PathBuf> {
+    get_sysroot_fast()
+        .ok_or(())
+        .or_else(|()| get_sysroot_slow())
+}
+
+fn passthrough_rustc(at_args: &[String], sysroot: &Path) -> eyre::Result<ExitStatus> {
+    // We can skip a `rustup` `rustc` -> real `rustc` resolution by just invoking the real `rustc` ourselves.
+    // TODO(kkysen) this might not work without `rustup`
+    let rustc = sysroot.join("bin/rustc");
```

Comment:
See other comment, but if cargo passes us a valid rustc path we should just use that rather than try to find the right one ourselves.

Reply:
Unfortunately not.  We can use `rustc --print sysroot`, though that's slower (subprocess) and harder to get right with non-UTF-8 paths.